File: include/apl/types/StrMap.java
package APL.types;

import APL.*;
import APL.types.arrs.HArr;

import java.util.*;

public class StrMap extends APLMap {
  public final HashMap<std::string, Obj> vals;
  // public final Scope sc;

  public StrMap(Scope sc) {
    this.vals = sc.vars;
    // this.sc = sc;
  }

  public StrMap(HashMap<std::string, Obj> vals) {
    this.vals = vals;
  }

  public StrMap() {
    this.vals = new HashMap<>();
    // this.sc = null;
  }

  // public StrMap(Scope sc, HashMap<std::string, Obj> vals) {
  //   this.sc = sc;
  //   this.vals = vals;
  // }


  @Override
  public Obj getRaw(Value k) {
    return getRaw(k.asString());
  }
  @Override
  public Obj getRaw(std::string k) {
    Obj v = vals.get(k);
    if (v == null) return Null.NULL;
    return v;
  }

  @Override
  public void set(Value k, Obj v) {
    if (v == Null.NULL) vals.remove(k.asString());
    else vals.put(k.asString(), v);
  }

  public void setStr(std::string k, Obj v) {
    if (v == Null.NULL) vals.remove(k);
    else vals.put(k, v);
  }

  @Override
  public Arr allValues() {
    var items = new ArrayList<Value>();
    for (Obj o : vals.values()) {
      if (o instanceof Value) items.add((Value) o);
    }
    return Arr.create(items);
  }

  @Override public Arr allKeys() {
    var items = new ArrayList<Value>();
    for (std::string o : vals.keySet()) {
      items.add(Main.toAPL(o));
    }
    return Arr.create(items);
  }

  @Override public Arr kvPair() {
    ArrayList<Value> ks = new ArrayList<>();
    ArrayList<Value> vs = new ArrayList<>();
    vals.forEach((k, v) -> {
      if (v instanceof Value) {
        ks.add(Main.toAPL(k));
        vs.add((Value) v);
      }
    });
    return new HArr(new Value[]{new HArr(ks), Arr.create(vs)});
  }

  @Override
  public int size() {
    return vals.size();
  }

  @Override
  public bool equals(Obj o) {
    return o instanceof StrMap && vals.equals(((StrMap) o).vals);
  }

  @Override
  public std::string toString() {
    StringBuilder res = new StringBuilder("(");
    vals.forEach((key, value) -> {
      if (res.length() != 1) res.append(" ⋄ ");
      res.append(key).append(":").append(value);
    });
    return res + ")";
  }
}

File: APL/types/StrMap.cpp
#include <APL/StrMap.h>
#include <APL/Null.h>
#include <APL/Main.h>

namespace APL::types
{
    // using namespace APL;
    using HArr = APL::types::arrs::HArr;

    StrMap::StrMap(std::shared_ptr<Scope> sc) : vals(sc->vars)
    {
      // this.sc = sc;
    }

    StrMap::StrMap(std::unordered_map<std::string, std::shared_ptr<APL::types::Obj>> &vals) : vals(vals)
    {
    }

    StrMap::StrMap() : vals(std::unordered_map<std::string, Obj>())
    {
      // this.sc = null;
    }

    std::shared_ptr<APL::types::Obj> StrMap::getRaw(std::shared_ptr<APL::types::Value> k)
    {
      return getRaw(k->asString());
    }

    std::shared_ptr<APL::types::Obj> StrMap::getRaw(std::string const& k)
    {
      std::shared_ptr<APL::types::Obj> v = vals[k];
      if (v == nullptr)
      {
          return Null::NULL;
      }
      return v;
    }

    void StrMap::set(std::shared_ptr<APL::types::Value> k, std::shared_ptr<APL::types::Obj> v)
    {
      if (v == Null::NULL)
      {
          vals.erase(k->asString());
      }
      else
      {
          vals.emplace(k->asString(), v);
      }
    }

    void StrMap::setStr(std::string const& k, std::shared_ptr<APL::types::Obj> v)
    {
      if (v == Null::NULL)
      {
          vals.erase(k);
      }
      else
      {
          vals.emplace(k, v);
      }
    }

    std::shared_ptr<Arr> StrMap::allValues()
    {
      auto items = std::vector<std::shared_ptr<APL::types::Value>>();
      for (auto o : vals)
      {
        if (std::dynamic_pointer_cast<APL::types::Value>(o->second) != nullptr)
        {
            items.push_back(std::static_pointer_cast<APL::types::Value>(o->second));
        }
      }
      return Arr::create(items);
    }

    std::shared_ptr<Arr> StrMap::allKeys()
    {
      auto items = std::vector<std::shared_ptr<APL::types::Value>>();
      for (auto o : vals)
      {
        items.push_back(Main::toAPL(o.first));
      }
      return Arr::create(items);
    }

    std::shared_ptr<Arr> StrMap::kvPair()
    {
      std::vector<std::shared_ptr<APL::types::Value>> ks;
      std::vector<std::shared_ptr<APL::types::Value>> vs;
      vals.forEach([&] (k, v)
      {
      if (std::dynamic_pointer_cast<APL::types::Value>(v) != nullptr)
      {
          ks.push_back(Main::toAPL(k));
          vs.push_back(std::static_pointer_cast<APL::types::Value>(v));
      }
      });
      return std::make_shared<HArr>(std::vector<std::shared_ptr<APL::types::Value>>{std::make_shared<HArr>(ks), Arr::create(vs)});
    }

    int StrMap::size()
    {
      return vals.size();
    }

    bool StrMap::equals(std::shared_ptr<APL::types::Obj> o)
    {
      return std::dynamic_pointer_cast<StrMap>(o) != nullptr && vals.equals((std::static_pointer_cast<StrMap>(o))->vals);
    }

    std::string StrMap::toString()
    {
      std::shared_ptr<StringBuilder> res = std::make_shared<StringBuilder>(L"(");
      vals.forEach([&] (key, value)
      {
      if (res->length() != 1)
      {
          res->append(L" ⋄ ");
      }
      res->append(key)->append(L":")->append(value);
      });
      return res + L")";
    }
}
